redbpf/src/lib.rs

```diff
@@ -415,10 +492,45 @@ impl KProbe {
 
                 ProbeAttachType::Return => perf::open_kretprobe_perf_event(fn_name, offset)?,
             };
-            perf::attach_perf_event(fd, pfd)
+            let ret = perf::attach_perf_event(fd, pfd);
+            if ret.is_ok() {
+                self.attachment_points.push(KProbeAttachmentPoint {
+                    fn_name: fn_name.to_owned(),
+                    offset,
+                    pfd,
+                });
+            } else {
+                libc::close(pfd);
+            }
+            ret
         }
     }
 
+    /// Detach the `kprobe` or `kretprobe`
+    ///
+    /// This method is not needed to be called manually because all attachment
+    /// points are detached and closed automatically when `KProbe` is
+    /// dropped. But this method provides a feature for detaching a bpf program
+    /// from kprobe event selectively.
+    ///
+    /// # Example
+    /// ```no_run
+    /// use redbpf::Module;
+    /// let mut module = Module::parse(&std::fs::read("file.elf").unwrap()).unwrap();
+    /// for kprobe in module.kprobes_mut() {
+    ///     kprobe.attach_kprobe(&kprobe.name(), 0).unwrap();
+    ///     // do some stuff...
+    ///     kprobe.detach_kprobe(&kprobe.name(), 0).unwrap();
+    /// }
+    /// ```
+    pub fn detach_kprobe(&mut self, fn_name: &str, offset: u64) -> Result<()> {
+        // bpf program is detached from perf event and the perf event is closed
+        // by dropping KProbeAttachmentPoint
+        self.attachment_points
+            .retain(|ap| ap.fn_name != fn_name || ap.offset != offset);
```

Comment:
Might be more readable to use the De Morgan's laws here as `!(ap.fn_name == fn_name && ap.offset == offset)`